client-js/main/client/firebase-backend-client.js

```diff
@@ -0,0 +1,351 @@
+    const subscriptionService = new FirebaseSubscriptionService(endpoint);
+
+    return new FirebaseBackendClient(endpoint, firebaseClient, requestFactory, subscriptionService);
+  }
+
+  /**
+   * @inheritDoc
+   * @return {BackendClient.Fetch<T>}
+   * @template <T>
+   */
+  _fetchOf(query) {
+    // noinspection JSValidateTypes A static member class type is not resolved properly.
+    return new FirebaseBackendClient.Fetch({of: query, using: this});
+  }
+
+  /**
+   * @inheritDoc
+   */
+  _subscribeToTopic(topic) {
+    return new Promise((resolve, reject) => {
+      const spineTopic = topic.raw();
+      this._endpoint.subscribeTo(spineTopic)
+        .then(subscription => {
+          const path = subscription.id.value;
+          const subscriptions = {add: null, remove: null, change: null};
+          const add = Observable.create(observer => {
+            subscriptions.add = this._firebase.onChildAdded(path, value => {
+              const message = topic.convert(value);
+              observer.next(message);
+            });
+          });
+          const change = Observable.create(observer => {
+            subscriptions.change = this._firebase.onChildChanged(path, value => {
+              const message = topic.convert(value);
+              observer.next(message);
+            });
+          });
+          const remove = Observable.create(observer => {
+            subscriptions.remove = this._firebase.onChildRemoved(path, value => {
+              const message = topic.convert(value);
+              observer.next(message);
+            });
+          });
+          const subscriptionProto = FirebaseBackendClient.subscriptionProto(path, topic);
+          const entitySubscription = new EntitySubscription({
+            unsubscribedBy: () => {
+              FirebaseBackendClient._tearDownSubscriptions(subscriptions);
+            },
+            withObservables: {add, change, remove},
+            forSubscription: subscriptionProto
+          });
+          resolve(entitySubscription.toObject());
+          this._subscriptionService.add(entitySubscription);
+        })
+        .catch(reject);
+    });
+  }
+
+  /**
+   * Unsubscribes the provided Firebase subscriptions.
+   *
+   * @param {{add: Subscription, remove: Subscription, change: Subscription}} subscriptions
+   * @private
+   */
+  static _tearDownSubscriptions(subscriptions) {
+    if (!subscriptions.add.closed) {
+      subscriptions.add.unsubscribe();
+    }
+    if (!subscriptions.remove.closed) {
+      subscriptions.remove.unsubscribe();
+    }
+    if (!subscriptions.change.closed) {
+      subscriptions.change.unsubscribe();
+    }
+  }
+
+  /**
+   * Creates a Protobuf `Subscription` instance to communicate with Spine server.
+   *
+   * @param {String} path a path to object which gets updated in Firebase
+   * @param {TypedTopic} topic a topic for which the Subscription gets updates
+   */
+  static subscriptionProto(path, topic) {
```

Comment:
Mentioning `Proto` is weird if that's a public API. The fact we are using proto messages is just an implementation detail.